readthedocs/api/v3/views.py

```diff
@@ -0,0 +1,320 @@
+import django_filters.rest_framework as filters
+from django.contrib.auth.models import User
+from django.utils.safestring import mark_safe
+from rest_flex_fields.views import FlexFieldsMixin
+from rest_framework.authentication import TokenAuthentication
+from rest_framework.decorators import action
+from rest_framework.metadata import SimpleMetadata
+from rest_framework.mixins import (
+    CreateModelMixin,
+    ListModelMixin,
+    RetrieveModelMixin,
+    UpdateModelMixin,
+)
+from rest_framework.pagination import LimitOffsetPagination
+from rest_framework.renderers import BrowsableAPIRenderer
+from rest_framework.response import Response
+from rest_framework.throttling import AnonRateThrottle, UserRateThrottle
+from rest_framework.viewsets import GenericViewSet
+from rest_framework_extensions.mixins import NestedViewSetMixin
+
+from readthedocs.builds.models import Build, Version
+from readthedocs.core.utils import trigger_build
+from readthedocs.projects.models import Project
+
+from .filters import BuildFilter, ProjectFilter, VersionFilter
+from .mixins import APIAuthMixin
+from .permissions import PublicDetailPrivateListing
+from .renderers import AlphabeticalSortedJSONRenderer
+from .serializers import (
+    BuildCreateSerializer,
+    BuildSerializer,
+    ProjectSerializer,
+    UserSerializer,
+    VersionSerializer,
+    VersionUpdateSerializer,
+)
+
+
+class APIv3Settings:
+
+    """
+    Django REST Framework settings for APIv3.
+
+    Override global DRF settings for APIv3 in particular. All ViewSet should
+    inherit from this class to share/apply the same settings all over the APIv3.
+
+    .. note::
+
+        The only settings used from ``settings.REST_FRAMEWORK`` is
+        ``DEFAULT_THROTTLE_RATES`` since it's not possible to define here.
+    """
+
+    # Using only ``TokenAuthentication`` for now, so we can give access to
+    # specific carefully selected users only
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (PublicDetailPrivateListing,)
+
+    pagination_class = LimitOffsetPagination
+    LimitOffsetPagination.default_limit = 10
+
+    renderer_classes = (AlphabeticalSortedJSONRenderer, BrowsableAPIRenderer)
+    throttle_classes = (UserRateThrottle, AnonRateThrottle)
+    filter_backends = (filters.DjangoFilterBackend,)
+    metadata_class = SimpleMetadata
+
+
+class ProjectsViewSet(APIv3Settings, APIAuthMixin, NestedViewSetMixin,
+                      FlexFieldsMixin, ListModelMixin, RetrieveModelMixin,
+                      GenericViewSet):
```

Comment:
Rather than mixing in `ListModelMixin`, `RetrieveModelMixin` and `GenericViewSet` couldn't we instead mixin `ReadOnlyModelViewSet`?

Reply:
Good point.

The only reason is that I didn't know that it existed :wink: . Upgrading...